src/opencl/opencl.cpp

```diff
@@ -0,0 +1,403 @@
+
+    return instances->c_pinned;
+}
+
+extern "C"
+cl_int sloth256_189_pinned_free(EncodeOpenCLInstances* instances) {
+    if (instances->pinned_pos == -1)
+        return (cl_int)2036;
+
+    EncodeOpenCLInstance* instance = instances->instances + instances->pinned_pos;
+
+    cl_int err;
+
+    err = clEnqueueUnmapMemObject(instance->cq[0], instances->pinned, (void*)instances->c_pinned, 0, NULL, NULL); OCL_ERR_CHECK(err, err);
+    instances->c_pinned = NULL;
+    err = clReleaseMemObject(instances->pinned); OCL_ERR_CHECK(err, err);
+    err = clFlush(instance->cq[0]); OCL_ERR_CHECK(err, err);
+    err = clFinish(instance->cq[0]); OCL_ERR_CHECK(err, err);
+
+    return CL_SUCCESS;
+}
+
+extern "C"
+EncodeOpenCLInstances* sloth256_189_opencl_init(cl_int& err) {
+    EncodeOpenCLInstances* instances = (EncodeOpenCLInstances*)malloc(sizeof(EncodeOpenCLInstances));
+
+    std::vector<cl_device_id> all_devices = getAllGPUs(err); OCL_ERR_CHECK(err, instances);
+    instances->num_instances = all_devices.size();
+    instances->instances = (EncodeOpenCLInstance*)malloc(sizeof(EncodeOpenCLInstance) * instances->num_instances);
+
+    size_t k = 0;
+    for (int platform_i = 0; platform_i < 3; platform_i++) {
+        Platform platform = (Platform)platform_i;
+        std::vector<cl_device_id> devices;
+        switch (platform) {
+            case NVIDIA:
+                devices = getAllNvidiaGPUs(err); if (err != 2027) OCL_ERR_CHECK(err, instances);
+                break;
+            case AMD:
+                devices = getAllAMDGPUs(err); if (err != 2028) OCL_ERR_CHECK(err, instances);
+                break;
+            case INTEL:
+                devices = getAllIntelGPUs(err); if (err != 2029) OCL_ERR_CHECK(err, instances);
+                break;
+        }
+        err = CL_SUCCESS;
+
+        for (size_t j = 0; j < devices.size(); j++) {
+            EncodeOpenCLInstance* instance = instances->instances + k;
+            instance->platform = platform;
+
+            instance->device = devices[j];
+
+            instance->context = clCreateContext(NULL, 1, &instance->device, NULL, NULL, &err); OCL_ERR_CHECK(err, instances);
+
+            std::string encode_path = "src/opencl";
+            std::string header_str = "";
+            if (instance->platform == NVIDIA) {
+                std::string ptx_str = read_from_file(encode_path + "/mod256-189.cu", err); OCL_ERR_CHECK(err, instances);
+                header_str = ptx_str + "\n" + read_from_file(encode_path + "/nvidia_specific.cl", err); OCL_ERR_CHECK(err, instances);
+            }
+            else {
+                header_str = read_from_file(encode_path + "/non_nvidia.cl", err); OCL_ERR_CHECK(err, instances);
+            }
+            std::string kernel_str = read_from_file(encode_path + "/encode.cl", err); OCL_ERR_CHECK(err, instances);

```

Comment:
I assume this will not work as files are moved around? Like executable using this library moved somewhere with no source code available anymore. If so, then source code of those should be bundled with executable (can be `include_bytes` macro in Rust and pointer sent into here for instance).